test/singledc/antidote_SUITE.erl

```diff
@@ -212,3 +218,138 @@ txn_seq_update_check(Node, TxId, Updates) ->
                       Res = rpc:call(Node, antidote, update_objects, [[Update], TxId]),
                       ?assertMatch(ok, Res)
               end, Updates).
+
+interactive_txn_validate_or_read(Config) ->
+    Bucket = ?BUCKET,
+    Node = proplists:get_value(node, Config),
+    Type = antidote_crdt_counter_pn,
+
+    Object1 = {vor_key1, Type, Bucket},
+
+    {ok, TxId1} = rpc:call(Node, antidote, start_transaction, [ignore, []]),
+
+    {ok, Res1} = rpc:call(Node, antidote, read_objects, [[Object1], TxId1]),
+    ?assertEqual([0], Res1),
+
+    % An empty token is invalid and returns the correct next token
+    {ok, Res2} = rpc:call(Node, antidote, validate_or_read_objects, [[Object1], [<<>>], TxId1]),
+    [{invalid, 0, Token1}] = Res2,
+
+    % With a correct token, no new value is returned.
+    {ok, Res3} = rpc:call(Node, antidote, validate_or_read_objects, [[Object1], [Token1], TxId1]),
+    ?assertEqual([valid], Res3),
+
+    {ok, Clock} = rpc:call(Node, antidote, commit_transaction, [TxId1]),
+    {ok, TxId2} = rpc:call(Node, antidote, start_transaction, [Clock, []]),
+
+    % Tokens works across transaction boundaries.
+    {ok, Res4} = rpc:call(Node, antidote, validate_or_read_objects, [[Object1], [Token1], TxId2]),
+    ?assertEqual([valid], Res4),
+
+    Update = {Object1, increment, 1},
+    ok = rpc:call(Node, antidote, update_objects, [[Update], TxId2]),
+
+    % Local updates discard the validate_or_read optim, token will always
+    % be invalid with this transaction.
+    {ok, Res5} = rpc:call(Node, antidote, validate_or_read_objects, [[Object1], [Token1], TxId2]),
+    ?assertEqual([{invalid, 1, <<>>}], Res5),
+
+    {ok, Res6} = rpc:call(Node, antidote, validate_or_read_objects, [[Object1], [<<>>], TxId2]),
+    ?assertEqual([{invalid, 1, <<>>}], Res6),
+
+    {ok, _} = rpc:call(Node, antidote, commit_transaction, [TxId2]).
+
+interactive_txn_validate_or_read_concurrent(Config) ->
+    Bucket = ?BUCKET,
+    Node = proplists:get_value(node, Config),
+    Type = antidote_crdt_counter_pn,
+
+    Object1 = {vor_key1, Type, Bucket},
+
+    {ok, TxIdA1} = rpc:call(Node, antidote, start_transaction, [ignore, []]),
+    {ok, TxIdB1} = rpc:call(Node, antidote, start_transaction, [ignore, []]),
+
+    % On the first read of an empty object without a min transaction time, there is no
+    % other choice but to return an invalid token.
```

Comment:
I am not sure of why this happens and if I can / should prevent it. If someone has some input on this...

Reply:
Can you add some comment to the definition of invalid token explaining its meaning?

Reply:
Done!